lighthouse-core/test/gather/gatherers/source-maps-test.js

```diff
@@ -0,0 +1,241 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-env jest */
+
+jest.useFakeTimers();
+
+const Driver = require('../../../gather/driver.js');
+const Connection = require('../../../gather/connections/connection.js');
+const SourceMaps = require('../../../gather/gatherers/source-maps.js');
+const {createMockSendCommandFn, createMockOnFn} = require('../mock-commands.js');
+
+const mapJson = JSON.stringify({
+  version: 3,
+  file: 'out.js',
+  sourceRoot: '',
+  sources: ['foo.js', 'bar.js'],
+  names: ['src', 'maps', 'are', 'fun'],
+  mappings: 'AAgBC,SAAQ,CAAEA',
+});
+
+describe('SourceMaps gatherer', () => {
+  /**
+   * `scriptParsedEvent` mocks the `sourceMapURL` and `url` seen from the protocol.
+   * `map` mocks the (JSON) of the source maps that `Runtime.evaluate` returns.
+   * `resolvedSourceMapUrl` is used to assert that the SourceMaps gatherer is using the expected
+   *                        url to fetch the source map.
+   * `fetchError` mocks an error that happens in the page. Only fetch error message make sense.
+   * @param {Array<{scriptParsedEvent: LH.Crdp.Debugger.ScriptParsedEvent, map: string, resolvedSourceMapUrl?: string, fetchError: string}>} mapsAndEvents
+   * @return {Promise<LH.Artifacts['SourceMaps']>}
+   */
+  async function runSourceMaps(mapsAndEvents) {
+    const onMock = createMockOnFn();
+
+    const sendCommandMock = createMockSendCommandFn()
+      .mockResponse('Debugger.enable', {})
+      .mockResponse('Debugger.disable', {});
+
+    for (const {scriptParsedEvent, map, resolvedSourceMapUrl, fetchError} of mapsAndEvents) {
+      onMock.mockEvent('protocolevent', {
+        method: 'Debugger.scriptParsed',
+        params: scriptParsedEvent,
+      });
+
+      if (scriptParsedEvent.sourceMapURL.startsWith('data:')) {
+        // Only the source maps that need to be fetched use the `evaluateAsync` code path.
+        continue;
+      }
+
+      if (map && fetchError) {
+        throw new Error('should only define map or fetchError, not both.');
+      }
+
+      sendCommandMock.mockResponse('Runtime.evaluate', ({expression}) => {
+        // Check that the source map url was resolved correctly. It'll be somewhere
+        // in the code sent to Runtime.evaluate.
+        if (resolvedSourceMapUrl && !expression.includes(resolvedSourceMapUrl)) {
+          throw new Error(`did not request expected url: ${resolvedSourceMapUrl}`);
+        }
+
+        const value = fetchError ?
+          Object.assign(new Error(), {message: fetchError, __failedInBrowser: true}) :
```

Comment:
ah, it's because of https://github.com/GoogleChrome/lighthouse/blob/21d00aca5d882add4386b9468b26dcaa64a0c28e/lighthouse-core/gather/driver.js#L541 (since `fetchError.message` isn't enumerable when message is set via the Error constructor, so it isn't copied over on that line)

fetchError could be a plain object like is done in the real `pageFunctions.wrapRuntimeEvalErrorInBrowserString`, but this seems good enough.